apps/evm/lib/evm/sub_state.ex

```diff
@@ -100,28 +98,27 @@ defmodule EVM.SubState do
 
   ## Examples
 
-      iex> substate1 = %EVM.SubState{refund: 5, logs: [1], selfdestruct_list: [5]}
-      iex> substate2 = %EVM.SubState{refund: 5, logs: [5], selfdestruct_list: [1]}
+      iex> substate1 = %EVM.SubState{refund: 5, logs: [1], selfdestruct_list: MapSet.new([5])}
+      iex> substate2 = %EVM.SubState{refund: 5, logs: [5], selfdestruct_list: MapSet.new([1])}
       iex> EVM.SubState.merge(substate1, substate2)
-      %EVM.SubState{refund: 10, logs: [1, 5], selfdestruct_list: [5, 1]}
+      %EVM.SubState{refund: 10, logs: [1, 5], selfdestruct_list: MapSet.new([5, 1])}
   """
 
   @spec merge(t(), t()) :: t()
   def merge(sub_state1, sub_state2) do
-    selfdestruct_list = sub_state1.selfdestruct_list ++ sub_state2.selfdestruct_list
-    dedup_selfdestruct_list = Enum.dedup(selfdestruct_list)
+    selfdestruct_list = MapSet.union(sub_state1.selfdestruct_list, sub_state2.selfdestruct_list)
     logs = sub_state1.logs ++ sub_state2.logs
 
     refund =
       sub_state1.refund + sub_state2.refund -
-        (Enum.count(selfdestruct_list) - Enum.count(dedup_selfdestruct_list)) *
-          Refunds.selfdestruct_refund()
+        (Enum.count(sub_state1.selfdestruct_list) + Enum.count(sub_state2.selfdestruct_list) -
+           Enum.count(selfdestruct_list)) * Refunds.selfdestruct_refund()
```

Comment:
I don't know the specification for how we calculate refunds, but this seems a bit odd to me. I'm guessing we should be doing the following: 

> Given sets `a` and `b`, the combined refund should be `refund a` + `refund b` - `refund of duplicate values`

Is that right? 

But we seem to be doing, 
```
a.refund + b.refund - ((a.count + b.count - (a union b).count) * refund_amount)
```
won't ` a.count + b.count` be the same as `(a union b).count`, making that `0 * refund_amount`? 

I think we might just want the _intersection_, right? 

```
a.refund + b.refund - ((a intersection b).count) * refund_amount)
```

Reply:
`a union b`  has only unique values from a and b. so `a.count + b.count != (a union b).count`. Your variant seem clearer so I will use it

Reply:
@germsvel done